jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/metadata/accessors/objects/MetadataAnnotatedElement.java

```diff
@@ -212,6 +212,9 @@ public MetadataAnnotation getAnnotation(String annotation) {
         MetadataAnnotation metadataAnnotation = m_annotations.get(annotation);
         if (metadataAnnotation == null) {
             for (MetadataAnnotation a: m_metaAnnotations.values()) {
+                if ( this.getName().equals( a.getName() ) ) { //Recursive
```

Comment:
Here we need more complex check. In A->B scenario _this_ is A and _a_ is B. If we walk trough A -> B -> C -> A, we have to make sure that a.getName() does not match _any_ this.getName() from the whole stack history.
I already made working solution with HashSet containing all the this.getName() values. But I don't like it much, it adds more complexity to he code.